## Investigasi sampel data titanic berikut dengan cara :
1. Cek secara head, tail, sample, info lalu observasi apa yang bisa anda peroleh ?
2. Lakukan Statistical Summary dengan mengekstrak informasi yang didapat dari observasi anda ?
3. Cek apakah ada duplikat dan bagaimana handlenya ?
4. Cek apakah ada missing value, berapa persentasenya jika ada, dan bagaimana cara handlenya ?

## Import Libraries

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

In [ ]:
# import data
df = pd.read_excel('titanic.xlsx')
df.head()

survived                                             name     sex      age
0         1                    Allen, Miss. Elisabeth Walton  female  29.0000
1         1                   Allison, Master. Hudson Trevor    male   0.9167
2         0                     Allison, Miss. Helen Loraine  female   2.0000
3         0             Allison, Mr. Hudson Joshua Creighton    male  30.0000
4         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000

In [ ]:
df.tail()

survived                                     name     sex   age
495         1  Mallet, Mrs. Albert (Antoinette Magnin)  female  24.0
496         0        Mangiavacchi, Mr. Serafino Emilio    male   NaN
497         0               Matthews, Mr. William John    male  30.0
498         0                Maybery, Mr. Frank Hubert    male  40.0
499         0                McCrae, Mr. Arthur Gordon    male  32.0

In [ ]:
df.sample(5)

survived                               name     sex   age
465         0         Jenkin, Mr. Stephen Curnow    male  32.0
259         1        Seward, Mr. Frederic Kimber    male  34.0
193         1  Madill, Miss. Georgette Alexandra  female  15.0
254         1              Saalfeld, Mr. Adolphe    male   NaN
26          1            Bishop, Mr. Dickinson H    male  25.0

Observation:


1. All columns except "age" are categorical.

2. The "survived" variable is binary (0,1).

3. The "sex" variable contains distinct values: "female" and "male".

4. There appears a variable with missing values, particularly in 'age'. This issue will be handle next

4. There are no significant issues found in the data (column names match the entries).




In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  500 non-null    int64  
 1   name      500 non-null    object 
 2   sex       500 non-null    object 
 3   age       451 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.8+ KB


1. The data contains 4 columns with 500 rows.  
2. Only the **"age"** variable has missing values, which will be processed further.  
3. The **"age"** data contains float values, and values below 1 will be considered as observations with an age of less than 1 year. Need a comfirmation from the source of dataset
4. Otherwise, the data appears to align with the column names.

#Statistical Summary

In [ ]:
df.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [ ]:
categoric = ['name', 'sex']
numeric = ['survived','age']

In [ ]:
df[numeric].describe()

survived         age
count  500.000000  451.000000
mean     0.540000   35.917775
std      0.498897   14.766454
min      0.000000    0.666700
25%      0.000000   24.000000
50%      1.000000   35.000000
75%      1.000000   47.000000
max      1.000000   80.000000

1. Overall, the minimum and maximum values seem reasonable for this variable.  
2. The **"age"** variable has a minimum value of 0.6, which can be considered as an age of less than 1 year.  
3. The mean is approximately 50% (Median), indicating that the variable is close to a normal distribution.  
4. The **"survived"** variable is binary (0 and 1), so there is no need to conclude its symmetry. It only needs to be checked for balance.

In [ ]:
df[categoric].describe()

name   sex
count                              500   500
unique                             499     2
top     Eustis, Miss. Elizabeth Mussey  male
freq                                 2   288

1. Both name and **sex** columns have 500 entries with no missing values.

2. There are 499 unique names, indicating one duplicate entry: "Eustis, Miss. Elizabeth Mussey", which suggests a potential duplicate that may need further checking.

3. The dataset contains 288 males and 212 females, showing that males are more represented.

In [ ]:
#untuk mengetahui berapa banyak gender dan status survived dengan detail
print(df['sex'].value_counts(), '\n')
print(df['survived'].value_counts())

sex
male      288
female    212
Name: count, dtype: int64 

survived
1    270
0    230
Name: count, dtype: int64




1.   For gender, it is true that "male" appears in more rows than "female".
2.   It shows that most passengers survived, with 270 survivors, while the rest did not.



In [ ]:
print(df['age'].value_counts())

age
24.0000    23
30.0000    20
36.0000    19
18.0000    14
42.0000    14
45.0000    14
35.0000    14
22.0000    12
28.0000    12
23.0000    11
21.0000    11
31.0000    11
48.0000    11
25.0000    11
29.0000    10
54.0000    10
50.0000    10
34.0000    10
47.0000    10
39.0000     9
27.0000     9
33.0000     8
49.0000     8
38.0000     8
19.0000     8
40.0000     8
55.0000     8
32.0000     7
60.0000     7
17.0000     7
26.0000     7
44.0000     6
52.0000     6
37.0000     6
58.0000     6
64.0000     5
46.0000     5
57.0000     5
51.0000     5
16.0000     4
53.0000     4
56.0000     4
43.0000     4
62.0000     4
61.0000     4
1.0000      3
8.0000      3
63.0000     3
41.0000     3
15.0000     2
71.0000     2
20.0000     2
65.0000     2
4.0000      2
6.0000      2
2.0000      1
0.9167      1
14.0000     1
11.0000     1
76.0000     1
59.0000     1
80.0000     1
28.5000     1
32.5000     1
70.0000     1
45.5000     1
13.0000     1
67.0000     1
12.0000     1
0.8333      1
0.6667      1
7.

The **"age"** variable shows that the most common age is 24 years, with 23 passengers, followed by 30 years, with 20 passengers.

#Cleaning Data

##Handling Duplicate

In [ ]:
len(df.drop_duplicates())/len(df)
#Hasil menunjukkan bahwa benar terdapat baris yang duplicate

0.998

In [ ]:
duplicates = df[df.duplicated(keep=False)]
duplicates
#Asumsi seelumnya benar, ada duplicate data pada baris 104 dan baris 349

survived                            name     sex   age
104         1  Eustis, Miss. Elizabeth Mussey  female  54.0
349         1  Eustis, Miss. Elizabeth Mussey  female  54.0

In [ ]:
#Handling Drop duplicate
data = df.drop_duplicates()
len(data.drop_duplicates())/len(data)

#====== Handling Duplicate Data Finish ==============

1.0

Handling data duplicate telah dilakukan

#Handling Missing Value

In [ ]:
data.isna().sum()

survived     0
name         0
sex          0
age         49
dtype: int64

In [ ]:
# percentage version
total_rows = len(data)

# Menghitung dan menampilkan persentase missing values di setiap kolom satu per satu
for column in data.columns:
    missing_count = data[column].isna().sum()
    missing_percentage = (missing_count / total_rows) * 100
    print(f"Column '{column}' Has {missing_count} missing values ({missing_percentage:.2f}%)") # .2f means 2 decimal

Column 'survived' Has 0 missing values (0.00%)
Column 'name' Has 0 missing values (0.00%)
Column 'sex' Has 0 missing values (0.00%)
Column 'age' Has 49 missing values (9.82%)


Terdapat 49 data missing value pada variabel **age**. Variabel lainnya telah lengkap.

Persentase missing value adalah 9.82% yang dimana berada dibawah 20%. Oleh karena itu, missing value ini dapat kita handling dengan **Median Imputation**, hal ini karena data **age** adalah data numerik

In [ ]:
data[column].fillna(data[column].median(), inplace=True)

<ipython-input-113-71226ee76d99>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data[column].fillna(data[column].median(), inplace=True)
<ipython-input-113-71226ee76d99>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[column].fillna(data[column].median(), inplace=True)


In [ ]:
data.isna().sum()

survived    0
name        0
sex         0
age         0
dtype: int64

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 499 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  499 non-null    int64  
 1   name      499 non-null    object 
 2   sex       499 non-null    object 
 3   age       499 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 19.5+ KB


**Duplicate** dan** Missing Values** pada data telah berhasil diselesaikan. Pada hasil akhir menunjukkan bahwa data yang bisa digunakan terdapat **499** baris **tanpa adanya missing value**.